[GLIMMER2] Remove remaining `CURRENT_TAG`s

chancancode opened the pull request:
![some context](https://cl.ly/grZn/Slack.png)

cc @krisselden @wycats

chancancode committed:
Remove unneeded mandatory-setter watcher

`REFERENCE_FOR_KEY` is only used for pooling property references, which
already handles the `mandatory-setter` check.

chancancode committed:
[GLIMMER2] Eliminate `PropertyReference`’s reliance on `CURRENT_TAG`

chancancode committed:
[GLIMMER2] Eliminate `ConditionalReference`’s reliance on `CURRENT_TAG`

This is simply not needed – `isTruthy` is a CP on the proxy that is only
updated when `content` changes.

rwjblue commented on packages/ember-runtime/lib/mixins/-proxy.js:
```diff
@@ -25,7 +25,7 @@ import symbol from 'ember-metal/symbol';
 
 export function isProxy(value) {
-  return value && value[IS_PROXY];
+  return typeof value === 'object' && value && value[IS_PROXY];
```
I think it may be better to do `typeof value === 'object' && value !== null` instead of coercing `value` to a boolean here.

chancancode replied:
I just memorized the dance @krisselden told me :P

rwjblue replied:
HAHA! Sounds good!

krisselden commented:
@rwjblue yes that is the only boolean coercion left,  it won't compile any others and the typeof prevents deopt on a different falsy coercion. The explicit check or the remaining coercion after the typeof check are the same at that point I lean toward the shorter one?

krisselden commented:
Likely though if isProxy is inlined into a hot path it will deopt on the first proxy it sees since they aren't very common anymore outside of ED promise proxy

chancancode committed:
[GLIMMER2] Refactor `{{#each-in}}`

* Make a specialized `Iterable` class for `{{#each-in}}`

* Handle/test duplicated keys in `{{#each-in}}`

* Default `{{#each-in}}` to use `key=“@index”` since it’s guaranteed to
  be unique (and is effectively free to compute)

* `ArrayIterable` doesn’t actually need to care about proxies: `ArrayProxy`
  is not actually a proxy – it does not inherit from `ProxyMixin`, thus
  does not contain the `[IS_PROXY]` brand. For tagging, since it listens
  for array change events on its `content` and update its `length`, the
  array proxy object already gets dirtied automatically whenever the
  content changes. For iterating, it falls into the `EmberArray` branch,
  which iterates through `arrangedContent`.

chancancode committed:
[GLIMMER2] Remove last `CURRENT_TAG` from `{{#each-in}}`

Basically copies the strategy from `PropertyReference`

chancancode commented:
The last commit removed all usage of `CURRENT_TAG`s 🎉

chancancode committed:
[GLIMMER2] Bump Glimmer for non-volatile iterables

chancancode commented:
This took care of the "big page re-rendering regression" on skylight (the problem was that we have a big list with ED promise proxies in each row)

chancancode committed:
[GLIMMER2] Only dirty `UpdatableReference` when value has changed

This path is used for Iterable’s `retain` operation. A lot of times, it
is called when nothing have actually changed. Dirtying the tag here in
those cases is causing a lot of unnecessary re-render churns.

krisselden commented:
If you bump a git npm dep like glimmer you need to clear caches on Travis or the blueprints test will fail.  Not sure why but it consistently fails with the cache.
